lighthouse-plugin-publisher-ads/audits/blocking-load-events.js

```diff
@@ -0,0 +1,231 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+const i18n = require('lighthouse/lighthouse-core/lib/i18n/i18n');
+const NetworkRecords = require('lighthouse/lighthouse-core/computed/network-records');
+// @ts-ignore
+const TraceOfTab = require('lighthouse/lighthouse-core/computed/trace-of-tab');
+const {Audit} = require('lighthouse');
+const {computeAdRequestWaterfall} = require('../utils/graph');
+const {getTimingsByRecord} = require('../utils/network-timing');
+
+/** @typedef {LH.Artifacts.NetworkRequest} NetworkRequest */
+/** @typedef {LH.Gatherer.Simulation.NodeTiming} NodeTiming */
+
+const UIStrings = {
+  title: 'Ads not blocked by load events',
+  failureTitle: 'Avoid waiting on load events',
+  description: 'Waiting on load events increases ad latency. ' +
+    'To speed up ads, eliminate the following load event handlers.',
+  displayValue: '{timeInMs, number, seconds} s blocked',
+  columnEvent: 'Event Name',
+  columnTime: 'Event Time',
+  columnScript: 'Script',
+  columnBlockedUrl: 'Blocked URL',
+  columnFunctionName: 'Function',
+};
+
+const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+
+/**
+ * Table headings for audits details sections.
+ * @type {LH.Audit.Details.Table['headings']}
+ */
+const HEADINGS = [
+  {key: 'eventName', itemType: 'text', text: str_(UIStrings.columnEvent)},
+  {key: 'time', itemType: 'ms', text: str_(UIStrings.columnTime), granularity: 1},
+  {key: 'url', itemType: 'url', text: str_(UIStrings.columnScript)},
+  {key: 'functionName', itemType: 'text', text: str_(UIStrings.columnFunctionName)},
+];
+
+/** @typedef {import('../utils/graph').SimpleRequest} SimpleRequest */
+
+/**
+ * @param {SimpleRequest} request
+ * @return {LH.Crdp.Runtime.CallFrame}
+ */
+function findOriginalCallFrame(request) {
+  let stack = request.initiator && request.initiator.stack;
+  if (!stack) {
+    return undefined;
+  }
+  while (stack.parent) {
+    stack = stack.parent;
+  }
+  return stack.callFrames[stack.callFrames.length - 1];
+}
+
+/**
+ * @param {LH.Crdp.Runtime.CallFrame} callFrame
+ * @param {LH.TraceEvent[]} traceEvents
+ */
+function findTraceEventOfCallFrame(callFrame, traceEvents) {
+  return traceEvents.find((e) => e.name == 'FunctionCall'
+      && e.args.data &&
+      // @ts-ignore
+      e.args.data.functionName == callFrame.functionName &&
+      // @ts-ignore
+      e.args.data.scriptId == callFrame.scriptId &&
+      // @ts-ignore
+      e.args.data.url == callFrame.url &&
+      // Tolerate off by one errors in line/column numbers.
+      // @ts-ignore
+      Math.abs(e.args.data.lineNumber == callFrame.lineNumber) < 2 &&
+      // @ts-ignore
+      Math.abs(e.args.data.columnNumber == callFrame.columnNumber) < 2);
```

Comment:
I'm somewhat inclined to suggest extracting the 2 to a const. But...probably YAGNI.